### PR TITLE
Opposite of progress CYOA

### PR DESCRIPTION
**Pull request by ddzz**
Good deeds feed CYOA.  I've defined a "good deed" as a bill that has been passed by Congress, signed by the President, and enacted.  Users can add/remove legislators from their favorites.  I did not use Devise; user login/authentication is handled through the Sessions controller/helper and the User model.

I've deployed to Heroku: https://quiet-hollows-7220.herokuapp.com/

and have created an account with the following credentials:
Email: ddzien@enova.com
Password: password

@jmmastey: I'd like to discuss this in person whenever you're free.

### CODE REVIEW

File: opposite_of_progress/good_deeds_feed/app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    redirect_to users_url
+  end
+
+  private
+
+  def user_params
+    params.require(:user).permit(:name, :email, :password, :password_confirmation)
+  end
+
+  def logged_in_user
```
Comment (djkotowski) [inline opposite_of_progress/good_deeds_feed/app/controllers/users_controller.rb]: Given you're using this method in multiple controllers, it makes more sense to include this in `ApplicationController` instead.

File: opposite_of_progress/good_deeds_feed/config/database.yml [hunk 2]
```diff
@@ -0,0 +1,88 @@
+# PostgreSQL. Versions 8.2 and up are supported.
```
Comment (djkotowski) [inline opposite_of_progress/good_deeds_feed/config/database.yml]: Get rid of unnecessary generated comments.

File: opposite_of_progress/good_deeds_feed/app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  end
+
+  def correct_user
+    @user = User.find(params[:id])
+    redirect_to(root_url) unless current_user?(@user)
+  end
+
+  def admin_user
+    redirect_to(root_url) unless current_user.admin?
+  end
```
Comment (djkotowski) [inline opposite_of_progress/good_deeds_feed/app/controllers/users_controller.rb]: The two methods above seem like they also belong in `ApplicationController` for reusability.

File: opposite_of_progress/good_deeds_feed/app/helpers/sessions_helper.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    if (user_id = session[:user_id])
+      @current_user ||= User.find_by(id: user_id)
+    elsif (user_id = cookies.signed[:user_id])
+      user = User.find_by(id: user_id)
+      if user && user.authenticated?(cookies[:remember_token])
+        log_in user
+        @current_user = user
+      end
+    end
+  end
```
Comment (djkotowski) [inline opposite_of_progress/good_deeds_feed/app/helpers/sessions_helper.rb]: This method is a little on the bulky side.  Try splitting it up a bit.

File: opposite_of_progress/good_deeds_feed/app/helpers/sessions_helper.rb [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+  def logged_in?
+    !current_user.nil?
```
Comment (djkotowski) [inline opposite_of_progress/good_deeds_feed/app/helpers/sessions_helper.rb]: Get rid of the negation.

```ruby
current_user.present?
```

File: opposite_of_progress/good_deeds_feed/app/models/good_deed.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  validates :congress_url, presence: true, uniqueness: { case_sensitive: false }
+  validates :short_title, presence: true, uniqueness: { case_sensitive: false }
+  validates :official_title,  presence: true,
+                              uniqueness: { case_sensitive: false }
+  validates :introduced_on, presence: true
+  validates :legislator_id, presence: true
+
+  def self.by_party(party)
+    joins(:legislator).where(legislators: { party: party })
+  end
```
Comment (djkotowski) [inline opposite_of_progress/good_deeds_feed/app/models/good_deed.rb]: Why not just make this a scope?

File: opposite_of_progress/good_deeds_feed/app/models/user.rb [hunk 7]
```diff
@@ -0,0 +1,48 @@
+class User < ActiveRecord::Base
+  attr_accessor :remember_token
+
+  has_many :favorites
+  has_many :legislators, through: :favorites
+  has_many :good_deeds, through: :legislators
+
+  before_save { self.email = email.downcase }
+  validates :name,  presence: true, length: { maximum: 50 }
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
```
Comment (djkotowski) [inline opposite_of_progress/good_deeds_feed/app/models/user.rb]: I'd move this elsewhere.  Having a constant definition in the middle of a group of validations is confusing, even if it is being used for the following validation.

File: opposite_of_progress/good_deeds_feed/app/models/user.rb [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  validates :email, presence: true, length: { maximum: 255 },
+                    format: { with: VALID_EMAIL_REGEX },
+                    uniqueness: { case_sensitive: false }
+  has_secure_password
```
Comment (djkotowski) [inline opposite_of_progress/good_deeds_feed/app/models/user.rb]: Likewise, this feels out of place.

File: opposite_of_progress/good_deeds_feed/db/seeds.rb [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+
+50.times do |n|
+  name  = Faker::Name.name
+  email = "example-#{n + 1}@gmail.com"
+  password = "password"
+  User.create!(name:  name,
+               email: email,
+               password:              password,
+               password_confirmation: password)
+end
```
Comment (djkotowski) [inline opposite_of_progress/good_deeds_feed/db/seeds.rb]: Save seeds for static production data.  FactoryGirl would be fine for this usage.

File: opposite_of_progress/good_deeds_feed/features/add_or_remove_favorites.feature [hunk 10]
```diff
@@ -0,0 +1,39 @@
+Feature: Add or remove legislator from favorites
+  As a user
+  I want to add or remove legislators from my favorites
+  To select which legislators will be included in my email updates
+
+  Background:
+    Given a legislator exists
+    Given a user exists and is logged in
```
Comment (djkotowski) [inline opposite_of_progress/good_deeds_feed/features/add_or_remove_favorites.feature]: `And a user exists and logged in`

File: opposite_of_progress/good_deeds_feed/features/add_or_remove_favorites.feature [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+  Scenario: Find button to add legislator to favorites
+    Given I am at a legislator's page
+    Then I see a button to add the legislator to my favorites
```
Comment (djkotowski) [inline opposite_of_progress/good_deeds_feed/features/add_or_remove_favorites.feature]: This test isn't necessary.  No need to have a test to find a button and then a test to use the button.

File: opposite_of_progress/good_deeds_feed/features/add_or_remove_favorites.feature [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+  Scenario: Remove legislator from user's profile page
+    Given the legislator is already in my favorites
+    And I am at my user profile page
+    When I remove the legislator from my favorites through my profile
+    Then the legislator is removed from my favorites
+
+  Scenario: Can't add to favorites if logged out
+    Given I have logged out of my account
+    When I am at a legislator's page
+    Then I do not see a button to add the legislator to my favorites
```
Comment (djkotowski) [inline opposite_of_progress/good_deeds_feed/features/add_or_remove_favorites.feature]: This test _is_ necessary, on the other hand.  Looks good.

File: opposite_of_progress/good_deeds_feed/app/controllers/good_deeds_controller.rb [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+    @good_deeds = GoodDeed.order(:legislator_id).paginate(page: params[:page])
+  end
+
+  def show
+    @good_deed = GoodDeed.find(params[:id])
+  end
+
+  def by_party
+    @good_deeds = GoodDeed.by_party(params[:party]).
+                    paginate(page: params[:page])
```
Comment (djkotowski) [inline opposite_of_progress/good_deeds_feed/app/controllers/good_deeds_controller.rb]: Odd indentation.

File: opposite_of_progress/good_deeds_feed/features/list_all_legislators.feature [hunk 14]
```diff
@@ -0,0 +1,18 @@
+Feature: Viewing all legislators
+  As a user
+  I want to see a list of all legislators
+  To see which legislators I can add to my favorites
+
+  Background:
+    Given many legislators exist
+    
+  Scenario: View first page of legislators
+    Given I am at the "legislators" page
```
Comment (djkotowski) [inline opposite_of_progress/good_deeds_feed/features/list_all_legislators.feature]: This is the action that leads to the `Then`.  Should really be a `When`.

File: opposite_of_progress/good_deeds_feed/features/step_definitions/list_all_legislators_steps.rb [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+Then(/^I see pagination links for the list of legislators$/) do
+  expect(page).to have_selector(".pagination")
+end
+
+Then(/^I see links to filter by party$/) do
+  expect(page).to have_selector("#party_filters")
+end
+
+Given(/^I navigate to the next page of legislators$/) do
+  visit("/legislators/index?page=2")
```
Comment (djkotowski) [inline opposite_of_progress/good_deeds_feed/features/step_definitions/list_all_legislators_steps.rb]: This is fragile as it assumes the next page is always 2.  I'd recommend using your UI (Capybara) for this navigation.

File: opposite_of_progress/good_deeds_feed/features/step_definitions/user_logs_in_or_out_steps.rb [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+end
+
+When(/^I try to log in without entering a username\/password$/) do
+  click_button("Log in")
+end
+
+Given(/^I have logged in to my account$/) do
+  fill_in("Email", with: @user.email)
+  fill_in("Password", with: @user.password)
+  click_button("Log in")
```
Comment (djkotowski) [inline opposite_of_progress/good_deeds_feed/features/step_definitions/user_logs_in_or_out_steps.rb]: Given you're reusing the same three steps, I recommend extracting the `fill_in`, `fill_in`, `click` logic to a method.

File: opposite_of_progress/good_deeds_feed/features/user_creates_account.feature [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+
+  Scenario: Create account with valid parameters
+    Given I have entered a valid name, email and password
+    When I try to create an account
+    Then my account is created
+
+  Scenario: Leave name blank
+    Given I have not entered a name
+    When I try to create an account
+    Then my account is not created and I see an error
```
Comment (djkotowski) [inline opposite_of_progress/good_deeds_feed/features/user_creates_account.feature]: This looks like two steps in one.

File: opposite_of_progress/good_deeds_feed/features/user_logs_in_or_out.feature [hunk 18]
```diff
@@ -0,0 +1,36 @@
+Feature: Logging in to or out of web account
+    As a user
+    I want to log in to or out my account
+    To see which legislators I've added to my favorites
+
+    Background:
+    Given a user account exists
+    And a legislator exists
+    And the legislator is in the user's favorites
```
Comment (djkotowski) [inline opposite_of_progress/good_deeds_feed/features/user_logs_in_or_out.feature]: Be consistent with naming.

```cucumber
And the legislator is in my favorites
```

File: opposite_of_progress/good_deeds_feed/lib/api_parser.rb [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+class CongressApiParser
+  include HTTParty
+
+  OK_STATUS_CODE = 200
+  BILLS_PER_PAGE = 50
+  BILL_KEYS = %w( congress urls official_title introduced_on sponsor_id
+                  short_title )
+
+  base_uri 'congress.api.sunlightfoundation.com'
+  default_params apikey: '23b3ee3083ea405dbb84c2b3476efcd6'
```
Comment (djkotowski) [inline opposite_of_progress/good_deeds_feed/lib/api_parser.rb]: Don't hard-code API keys.  This gives anyone looking at your code access to your key.

File: opposite_of_progress/good_deeds_feed/spec/models/good_deed_spec.rb [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+
+RSpec.describe GoodDeed, type: :model do
+  describe "#by_party" do
+    it "filters good deeds by party" do
+      D_deed = FactoryGirl.create(:good_deed, legislator:
+                FactoryGirl.create(:legislator, party: "D"))
+      R_deed = FactoryGirl.create(:good_deed, legislator:
+                FactoryGirl.create(:legislator, party: "R"))
+      I_deed = FactoryGirl.create(:good_deed, legislator:
+                FactoryGirl.create(:legislator, party: "I"))
```
Comment (djkotowski) [inline opposite_of_progress/good_deeds_feed/spec/models/good_deed_spec.rb]: For clarity, I recommend separating the legislator create and deed create steps.

File: opposite_of_progress/good_deeds_feed/lib/api_parser.rb (hunk 19 above)
Comment (ddzz) [inline opposite_of_progress/good_deeds_feed/lib/api_parser.rb]: Where should I store/enter the API key?

File: opposite_of_progress/good_deeds_feed/spec/models/good_deed_spec.rb [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+
+  describe "#to_json" do
+    it "returns good deed data in JSON format" do
+      FactoryGirl.create_list(:good_deed, 5)
+      json_deeds = GoodDeed.to_json
+      expect { JSON.parse(json_deeds) }.not_to raise_error
+      expect(JSON.parse(json_deeds).length).to eq(5)
+    end
+  end
+end
```
Comment (djkotowski) [inline opposite_of_progress/good_deeds_feed/spec/models/good_deed_spec.rb]: You may want to test `#send_email_notification` to ensure it calls GoodDeedsMailer.

File: opposite_of_progress/good_deeds_feed/spec/models/user_spec.rb [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+  end
+
+  describe "#forget" do
+    it "deletes the user's remember token" do
+      user.remember
+      user.forget
+      expect(user.remember_digest).to eq(nil)
+    end
+  end
+end
```
Comment (djkotowski) [inline opposite_of_progress/good_deeds_feed/spec/models/user_spec.rb]: I would recommend testing your custom email validation.

File: opposite_of_progress/good_deeds_feed/config/routes.rb [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+  post 'login' => 'sessions#create'
+  delete 'logout' => 'sessions#destroy'
+  get 'legislators_by_party' => 'legislators#by_party'
+  get 'good_deeds_by_party' => 'good_deeds#by_party'
+  get 'good_deeds_json_feed' => 'good_deeds#json_feed'
+  resources :legislators
+  resources :good_deeds
+  resources :users
+  resources :favorites, only: [:create, :destroy]
+end
```
Comment (djkotowski) [inline opposite_of_progress/good_deeds_feed/config/routes.rb]: I'd organize this a bit more.  As it stands, you have a lot of mixed paths.

**Comment by djkotowski** [comment]
You have leftover Minitest files from your app generation.

**Comment by ashtonthomas** [comment]
@ddzz, I've been wanting to take a look at this for the past week. I just cloned the project and set it up locally, and I plan on going through it tomorrow. Maybe we can get together over the next two days and go through any comments. Sorry for the delay!

**Comment by djkotowski** [comment]
This looks good.  Look at the comments, but there is no need for a resubmission.  Good work!

**Pull request closed without merging**